eslint.config.mjs

```diff
@@ -0,0 +1,416 @@
+  },
+
+  // Next.js entry points pages
+  {
+    files: ['docs/pages/**/*'],
+    rules: {
+      'react/prop-types': 'off',
+    },
+  },
+  // Common config from core end
+
+  {
+    files: [`docs/**/*.${EXTENSION_TS}`],
+    ignores: ['**/*.d.ts'],
+    rules: {
+      'no-restricted-imports': [
+        'error',
+        {
+          paths: RESTRICTED_TOP_LEVEL_IMPORTS.map((name) => ({
+            name,
+            message: 'Use deeper import instead',
+          })),
+          patterns: [
+            {
+              group: [
+                '@mui/*/*/*',
+                // Allow any import depth with any internal packages
+                '!@mui/internal-*/**',
+
+                // The scheduler import strategy is not determined yet
+                '!@mui/x-scheduler/**',
+
+                // Exceptions (QUESTION: Keep or remove?)
+                '!@mui/x-data-grid/internals/demo',
+                '!@mui/x-date-pickers/internals/demo',
+                '!@mui/x-tree-view/hooks/useTreeViewApiRef',
+                // TODO: export this from /ButtonBase in core. This will break after we move to package exports
+                '!@mui/material/ButtonBase/TouchRipple',
+              ],
+              message: 'Use less deep import instead',
+            },
+          ],
+        },
+      ],
+    },
+  },
+
+  {
+    files: ['packages/x-telemetry/**/*{.tsx,.ts,.js}'],
+    rules: {
+      'no-console': 'off',
+    },
+  },
+  {
+    files: ['packages/x-scheduler/**/*{.tsx,.ts,.js}'],
+    rules: {
+      // Base UI lint rules
+      '@typescript-eslint/no-redeclare': 'off',
+      'import/export': 'off',
+      'material-ui/straight-quotes': 'off',
+      'jsdoc/require-param': 'off',
+      'jsdoc/require-returns': 'off',
+    },
+  },
+  ...buildPackageRestrictedImports('@mui/x-charts', 'x-charts', false),
+  ...buildPackageRestrictedImports('@mui/x-charts-pro', 'x-charts-pro', false),
+  ...buildPackageRestrictedImports('@mui/x-charts-premium', 'x-charts-premium', false),
+  ...buildPackageRestrictedImports('@mui/x-codemod', 'x-codemod', false),
+  ...buildPackageRestrictedImports('@mui/x-data-grid', 'x-data-grid'),
+  ...buildPackageRestrictedImports('@mui/x-data-grid-pro', 'x-data-grid-pro'),
+  ...buildPackageRestrictedImports('@mui/x-data-grid-premium', 'x-data-grid-premium'),
+  ...buildPackageRestrictedImports('@mui/x-data-grid-generator', 'x-data-grid-generator'),
+  ...buildPackageRestrictedImports('@mui/x-date-pickers', 'x-date-pickers', false),
+  ...buildPackageRestrictedImports('@mui/x-date-pickers-pro', 'x-date-pickers-pro', false),
+  ...buildPackageRestrictedImports('@mui/x-tree-view', 'x-tree-view', false),
+  ...buildPackageRestrictedImports('@mui/x-tree-view-pro', 'x-tree-view-pro', false),
+  ...buildPackageRestrictedImports('@mui/x-license', 'x-license'),
+  ...buildPackageRestrictedImports('@mui/x-telemetry', 'x-telemetry'),
+
+  ...addReactCompilerRule(CHARTS_PACKAGES, ENABLE_REACT_COMPILER_PLUGIN_CHARTS),
```

Comment:
Nitpick, but in general, if possible, I'd avoid building rules with `files` key baked in. It makes it quite hard to reason about this config to not see the files it applies to in the config object. It could make sense to build the files array instead:

```tsx
  {
    files: Object.entries({
      [`packages/x-charts/src/**/*.${EXTENSION_TS}`]: ENABLE_REACT_COMPILER_PLUGIN_CHARTS,
      [`packages/x-data-grid/src/**/*.${EXTENSION_TS}`]: ENABLE_REACT_COMPILER_PLUGIN_DATA_GRID,
    }).reduce((acc, [path, enabled]) => (enabled ? [...acc, path] : acc), []),
    rules: {
      'react-compiler/react-compiler': 'error',
    },
  },
```

Reply:
Great idea. 👍 
But it can probably be kept for a follow-up, since this PR only refactors ESLint to FLAT config with little to no changes in the rules themselves. 🙈